stormpath/stormpath-sdk-node
# Refactor: Configuration

Changes the way that configuration is loaded by using the (new) [stormpath-config](https://github.com/stormpath/stormpath-node-config) module and its strategies.

This new way of loading configuration breaks the loading phase down into several configurable steps. Not only is this easier to test and manage. This also makes the configuration process extendable from other modules (see [express-stormpath branch refactor-client-config](https://github.com/stormpath/stormpath-express/tree/refactor-client-config)).

#### What this actually fixes

* **YAML supported.** There can now be both `~/.stormpath/stormpath.[json/yml]` and `./stormpath.[json/yml]` files. Also the default configuration (previously config.json) is now all YAML.
* **Fully async.** Previously the configuration was both sync (blocking fs calls) and async. With most of the configuration loaded sync during Client instantiation and some enrichment running after async and eventually triggering the 'ready' event. This made the client a bit inconsistent since one could both receive an exception and then errors through the 'error' event. Also, the client could be instantiated and used without having emitted its 'ready' event. Now, the 'ready' event will first be emitted once the configuration is fully loaded. Though, in order for users to still be able to just instantiate the client and call the API without waiting for the 'ready' event the client will be wrapped in a proxy and keep all calls to the Client waiting until the client is actually ready.

* **Only client config.** Previously the SDK contained configuration related to the express integration. This has now been moved into the [express-stormpath](https://github.com/stormpath/stormpath-express/blob/refactor-client-config/lib/client.js) module.

* **Testable.** Previously the configuration was a bit shattered and hard to test. Now it's separated into isolated strategies so testing can be done both separately (single strategy) and many together (with ConfigLoader, grouping of strategies).

(additional)

* **'Ready' event only fired once.** Previously the ready event was only fired once. So if someone attached to the client too late (after the event had been emitted), then they would never receive it. This is now fixed, so that if one attach to the 'ready' event, and the client is ready, then the callback is called immediately.

#### How to verify

1. Check out the `refactor-config-loader` branch.
2. Run `$ npm test` and hope for the best!
3. Run `npm link` to link this branch globally so that it's used when the `stormpath` module is referenced.
4. Create a new empty project (new directory).
5. Run `npm link stormpath` so that this project links any `stormpath` includes to our `refactor-config-loader` branch.
4. Remove any Stormpath environment variables.
3. Create a new file (e.g. `app.js`) and initialize a new stormath Client without providing any configuration (i.e. `new stormpath.Client()`).
4. Add a call to the `client.getApplications` method.
5. Run the application and verify that you get an error stating `Error: API key ID and secret is required`.
6. Create a `stormpath.(json or yml)` file in your `.stormpath` home folder. Add your Stormpath credentials.
7. Run the application and verify that you're able to retrieve the account.
8. Create a `stormpath.(json or yml)` file in the same directory as your application. Add some configuration that overwrites the configuration you specified in your home directory (step 3).
9. Run the application and verify that the new configuration you set has affected your client.
10. Add configuration to your client code (i.e. `new stormpath.Client({/* config */})`).
11. Run the application and verify that the configuration has overridden the previous configurations set (in step 3 and 8).

(optional/additional)

If you already have an existing application, then upgrade to the new `stormpath` module, then test the application. It should function exactly the same as before (except from the `express-stormpath` module, since this will have to be updated to work with the new configuration loader. See [stormpath-express/refactor-client-config](https://github.com/stormpath/stormpath-express/tree/refactor-client-config)).

(optional/additional)

Create a new application according to [the SDK tutorial](http://docs.stormpath.com/nodejs/quickstart/) and test that the application works as previously expected.
